go/cli/commands/build.go

```diff
@@ -30,80 +30,121 @@ import (
+	r.Command.Flags().StringVarP(&r.Docker.DockerfilePath, "dockerfile", "f", "",
+		"path to the Dockerfile. If not given, using a default builtin Dockerfile")
+	r.Command.Flags().StringVarP(&r.Ko.Repo, "repo", "r", "",
+		"the image repo. default to ko.local")
+	r.Command.Flags().StringVarP(&r.Ko.Tag, "tag", "t", "latest",
+		"the ko image tag")
+	// TODO: Docker CLI uses `--tag` flag to refer to "image:tag", which could be confusing but broadly accepted.
+	// We should better guide users on how to use "tag" and "image" flags for kfn.
+	// Here we use "tag" for ko <tag> (same as `ko build --tag`) and "image" for docker <image:tag> (same as `docker build --tag`)
+	return r
 }
 
 type BuildRunner struct {
 	ctx     context.Context
 	Command *cobra.Command
 
-	Tag            string
+	BuilderType string
+	Tag         string
+	Ko          *KoBuilder
+	Docker      *DockerBuilder
+}
+
+type Builder interface {
+	Build() error
+	Validate() error
+}
+
+type DockerBuilder struct {
+	Image          string
 	DockerfilePath string
 }
 
-func (r *BuildRunner) PreRunE(cmd *cobra.Command, args []string) error {
-	if err := r.requireDocker(); err != nil {
-		return err
-	}
-	if !r.dockerfileExist() {
-		err := r.createDockerfile()
-		if err != nil {
-			return err
-		}
-	}
-	return nil
+type KoBuilder struct {
+	Repo string
+	Tag  string
 }
 
 func (r *BuildRunner) RunE(cmd *cobra.Command, args []string) error {
-	return r.runDockerBuild()
+	var builder Builder
+	switch r.BuilderType {
+	case Docker:
+		builder = r.Docker
+	case Ko:
+		builder = r.Ko
+	}
+	if err := builder.Validate(); err != nil {
+		return err
+	}
+	return builder.Build()
 }
 
-func (r *BuildRunner) runDockerBuild() error {
-	args := []string{"build", ".", "-f", r.DockerfilePath, "--tag", r.Tag}
-	cmd := exec.Command("docker", args...)
-	var out, errout bytes.Buffer
-	cmd.Stdout = &out
-	cmd.Stderr = &errout
-	err := cmd.Run()
+func (r *DockerBuilder) Build() error {
+	args := []string{"build", ".", "-f", r.DockerfilePath, "--tag", r.Image}
+	out, err := execCmdFn(nil, "docker", args...)
 	if err != nil {
-		color.Red(strings.TrimSpace(errout.String()))
 		return err
 	}
-	color.Green(out.String())
-	color.Green("Image %v builds successfully. Now you can publish the image", r.Tag)
+	fmt.Println(out)
+	color.Green("Image %v builds successfully. Now you can publish the image", r.Image)
 	return nil
 }
 
-func (r *BuildRunner) requireDocker() error {
-	_, err := exec.LookPath("docker")
+func (r *DockerBuilder) Validate() error {
+	if err := r.validateDockerInstalled(); err != nil {
+		return err
+	}
+	if r.fileExist() {
+		return nil
+	}
+	return r.createDockerfile()
+}
+
+func (r *DockerBuilder) validateDockerInstalled() error {
+	_, err := execLookPathFn("docker")
 	if err != nil {
 		return fmt.Errorf("kfn requires that `docker` is installed and on the PATH")
 	}
 	return nil
 }
 
-func (r *BuildRunner) dockerfileExist() bool {
+func (r *DockerBuilder) fileExist() bool {
```

Comment:
Thanks! changed back to `dockerfileExists`.
Because I changed the object from `BuildRunner` to `DockerBuilder`, I thought it is verbose to repeat `docker`. similar to `docker build --file`.